presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -1037,9 +1044,12 @@ public PlanWithProperties visitIndexSource(IndexSourceNode node, PreferredProper
                             .build());
         }
 
-        private Function<VariableReferenceExpression, Optional<VariableReferenceExpression>> outputToInputTranslator(UnionNode node, int sourceIndex, TypeProvider types)
+        private Function<RowExpression, Optional<RowExpression>> outputToInputTranslator(UnionNode node, int sourceIndex)
         {
-            return variable -> Optional.of(node.getVariableMapping().get(variable).get(sourceIndex));
+            return variable -> {
+                checkArgument(variable instanceof VariableReferenceExpression, "Expect VariableReferenceExpression");
```

Comment:
nit: you can remove it. (VariableReferenceExpression) input will do the job and print the good error message